From: Ben Hutchings <ben@decadent.org.uk>
Date: Thu, 19 Aug 2010 02:55:33 +0100
Subject: [PATCH] i915: Blacklist i830, i845, i855 for KMS

Based on work by Andy Whitcroft <apw@canonical.com>.

We need to support partial upgrades in Debian so we require the X
driver to explicitly enable KMS by setting module parameter modeset=1.
Therefore this does not provide a means to override the blacklist.
---
 drivers/gpu/drm/i915/i915_drv.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -544,6 +544,15 @@ static struct drm_driver driver = {
 	.patchlevel = DRIVER_PATCHLEVEL,
 };
 
+/* KMS is reported to be unreliable for these devices */
+static const struct pci_device_id i915_kms_blacklist[] = {
+	INTEL_VGA_DEVICE(0x3577, 0),	/* i830 */
+	INTEL_VGA_DEVICE(0x2562, 0),	/* i845 */
+	INTEL_VGA_DEVICE(0x3582, 0),	/* i855 */
+	INTEL_VGA_DEVICE(0x358e, 0),	/* i855 */
+	{ },
+};
+
 static int __init i915_init(void)
 {
 	driver.num_ioctls = i915_max_ioctl;
@@ -566,6 +575,12 @@ static int __init i915_init(void)
 	if (i915_modeset == 1)
 		driver.driver_features |= DRIVER_MODESET;
 
+	if ((driver.driver_features & DRIVER_MODESET) &&
+	    pci_dev_present(i915_kms_blacklist)) {
+		DRM_INFO("KMS is not supported on this GPU\n");
+		driver.driver_features &= ~DRIVER_MODESET;
+	}
+
 #ifdef CONFIG_VGA_CONSOLE
 	if (vgacon_text_force() && i915_modeset == -1)
 		driver.driver_features &= ~DRIVER_MODESET;
